crates/primitives/src/epoch.rs

```diff
@@ -0,0 +1,109 @@
+//! Types relating to epoch bookkeeping.
+//!
+//! An epoch of a range of sequential blocks defined by the terminal block of
+//! the epoch going back to (but not including) the terminal block of a previous
+//! epoch.  This uniquely identifies the epoch's final state indirectly,
+//! although it's possible for conflicting epochs with different terminal blocks
+//! to exist in theory, depending on the consensus algorithm.
+//!
+//! Epochs are *usually* always the same number of slots, but we're not
+//! guaranteeing this yet, so we always include both the epoch number and slot
+//! number of the terminal block.
+//!
+//! We also have a sentinel "null" epoch used to refer to the "finalized epoch"
+//! as of the genesis block.
+
+use arbitrary::Arbitrary;
+use borsh::{BorshDeserialize, BorshSerialize};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    buf::Buf32,
+    l2::{L2BlockCommitment, L2BlockId},
+};
+
+/// Commits to a particular epoch by the last block and slot.
+#[derive(
+    Copy,
+    Clone,
+    Debug,
+    Eq,
+    PartialEq,
+    Ord,
+    PartialOrd,
+    Hash,
+    Arbitrary,
+    BorshDeserialize,
+    BorshSerialize,
+    Deserialize,
+    Serialize,
+)]
+pub struct EpochCommitment {
+    epoch: u64,
+    last_slot: u64,
+    last_blkid: L2BlockId,
+}
+
+impl EpochCommitment {
+    pub fn new(epoch: u64, last_slot: u64, last_blkid: L2BlockId) -> Self {
+        Self {
+            epoch,
+            last_slot,
+            last_blkid,
+        }
+    }
+
+    /// Creates a "null" epoch with
+    pub fn null() -> Self {
+        Self::new(0, 0, L2BlockId::from(Buf32::zero()))
+    }
+
+    pub fn epoch(&self) -> u64 {
+        self.epoch
+    }
+
+    pub fn last_slot(&self) -> u64 {
+        self.last_slot
+    }
+
+    pub fn last_blkid(&self) -> &L2BlockId {
+        &self.last_blkid
+    }
+
+    /// Returns a [`L2BlockCommitment`] for the final block of the epoch.
+    pub fn to_block_commitment(&self) -> L2BlockCommitment {
+        L2BlockCommitment::new(self.last_slot, self.last_blkid)
+    }
+
+    /// Returns if the terminal blkid is zero.
+    pub fn is_null(&self) -> bool {
+        Buf32::from(self.last_blkid).is_zero()
+    }
+
+    /// Checks if the epoch is sane.
+    ///
+    /// Ie. if the terminal blkid is zero then the last slot and epoch number
+    /// are also zero.
+    fn sanity_check(&self) -> bool {
+        if self.is_null() {
+            self.last_slot == 0 && self.epoch == 0
+        } else {
+            self.last_slot != 0
+        }
+    }
+}
```

Comment:
Why not include this check in the constructor?

Reply:
Maybe I should just get rid of the sanity check altogether.  This special casing isn't well-defined at the moment since it's not actually used in this context.

Reply:
Would it work to avoid the sentinel value and use an `Option` instead? Or does that screw up the commitment call...

Reply:
It'd be nice to always have a well-defined finalized block so that we don't have to do special casing just for the genesis.  Maybe we can rework the interfaces/definitions around this.